File: 3534. path-existence-queries-in-a-graph-ii.py
from typing import List


class Solution:
    def pathExistenceQueries(self, n: int, nums: List[int], maxDiff: int, queries: List[List[int]]) -> List[int]:
        new_nums = sorted(enumerate(nums), key=lambda x: x[1])
        get_i = [0] * n
        for i, (orig, _) in enumerate(new_nums):
            get_i[orig] = i

        LOG = 18
        st = [[0] * LOG for _ in range(n)]

        r = 0
        for i in range(n):
            if r < i:
                r = i
            while (
                r + 1 < n
                and new_nums[r + 1][1] - new_nums[r][1] <= maxDiff
                and new_nums[r + 1][1] - new_nums[i][1] <= maxDiff
            ):
                r += 1
            st[i][0] = r

        for j in range(1, LOG):
            for i in range(n):
                st[i][j] = st[st[i][j - 1]][j - 1]

        ans = []
        for u, v in queries:
            a, b = get_i[u], get_i[v]
            if a > b:
                a, b = b, a
            if a == b:
                ans.append(0)
                continue

            curr, steps = a, 0
            for j in range(LOG - 1, -1, -1):
                if st[curr][j] < b:
                    curr = st[curr][j]
                    steps += (1 << j)

            ans.append(steps + 1 if st[curr][0] >= b else -1)

        return ans


if __name__ == "__main__":
    sol = Solution()

    # Example 1
    n = 5
    nums = [1, 8, 3, 4, 2]
    maxDiff = 3
    queries = [[0, 3], [2, 4]]
    print(sol.pathExistenceQueries(n, nums, maxDiff, queries))  # [1, 1]

    # Example 2
    n = 5
    nums = [5, 3, 1, 9, 10]
    maxDiff = 2
    queries = [[0, 1], [0, 2], [2, 3], [4, 3]]
    print(sol.pathExistenceQueries(n, nums, maxDiff, queries))  # [1, 2, -1, 1]

    # Example 3
    n = 3
    nums = [3, 6, 1]
    maxDiff = 1
    queries = [[0, 0], [0, 1], [1, 2]]
    print(sol.pathExistenceQueries(n, nums, maxDiff, queries))  # [0, -1, -1]
}

File: 3534. path-existence-queries-in-a-graph-ii.java
public import java.util.Arrays;

class Solution {
    public int[] pathExistenceQueries(int n, int[] nums, int maxDiff, int[][] queries) {
        int[][] newNums = new int[n][2];
        for (int i = 0; i < n; i++)
            newNums[i] = new int[]{nums[i], i};

        Arrays.sort(newNums, (a, b) -> a[0] - b[0]);

        int[] getI = new int[n];
        for (int i = 0; i < n; i++)
            getI[newNums[i][1]] = i;

        int[][] st = new int[n][18];
        int r = 0;
        for (int i = 0; i < n; i++) {
            if (r < i)
                r = i;
            while (r + 1 < n &&
                   newNums[r + 1][0] - newNums[r][0] <= maxDiff &&
                   newNums[r + 1][0] - newNums[i][0] <= maxDiff)
                r++;
            st[i][0] = r;
        }

        for (int j = 1; j < 18; j++)
            for (int i = 0; i < n; i++)
                st[i][j] = st[st[i][j - 1]][j - 1];

        int[] ans = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            int a = getI[queries[i][0]], b = getI[queries[i][1]];
            if (a > b) {
                int t = a;
                a = b;
                b = t;
            }
            if (a == b) {
                ans[i] = 0;
                continue;
            }

            int curr = a, steps = 0;
            for (int j = 17; j >= 0; j--)
                if (st[curr][j] < b) {
                    curr = st[curr][j];
                    steps += (1 << j);
                }

            ans[i] = (st[curr][0] >= b) ? steps + 1 : -1;
        }

        return ans;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        // Example 1
        int n1 = 5;
        int[] nums1 = {1, 8, 3, 4, 2};
        int maxDiff1 = 3;
        int[][] queries1 = {{0, 3}, {2, 4}};
        System.out.println(Arrays.toString(sol.pathExistenceQueries(n1, nums1, maxDiff1, queries1))); // [1, 1]

        // Example 2
        int n2 = 5;
        int[] nums2 = {5, 3, 1, 9, 10};
        int maxDiff2 = 2;
        int[][] queries2 = {{0, 1}, {0, 2}, {2, 3}, {4, 3}};
        System.out.println(Arrays.toString(sol.pathExistenceQueries(n2, nums2, maxDiff2, queries2))); // [1, 2, -1, 1]

        // Example 3
        int n3 = 3;
        int[] nums3 = {3, 6, 1};
        int maxDiff3 = 1;
        int[][] queries3 = {{0, 0}, {0, 1}, {1, 2}};
        System.out.println(Arrays.toString(sol.pathExistenceQueries(n3, nums3, maxDiff3, queries3))); // [0, -1, -1]
    }
} {
	
}
